### PR TITLE
[IMP] pos_*: payment terminals in point of sale

### PR DESCRIPTION
**Pull request by paan-odoo**
pos_*: point_of_sale, pos_adyen, pos_mercado_pago, pos_paytm, pos_razorpay, pos_restaurant_adyen, pos_self_order_adyen, pos_six, pos_stripe, pos_viva_wallet, pos_mercury

Before this commit:
===========
- Previously, there was no model for storing mandatory credentials for different terminal methods.
- So, it was necessary to enter the same credentials with different terminal IDs for various payment terminal methods provided by the same vendor.
- This resulted in the duplication of credentials.

After this commit:
===========
- So, we are introducing a new 'pos.payment.provider' model for storing the same mandatory credentials in one model.
- We allow credentials to be entered once and used across multiple terminal payment methods. This enhancement streamlines the function of payment terminals and saves time during the creation of new payment methods.

Related PR:
- Enterprise: https://github.com/odoo/enterprise/pull/67165
- Upgrade: https://github.com/odoo/upgrade/pull/6248

task-3506646

### CODE REVIEW

File: addons/pos_mercado_pago/models/pos_payment_method.py [hunk 1]
```diff
@@ -78,12 +49,20 @@ def mp_payment_intent_cancel(self, payment_intent_id):
...
-        mercado_pago = MercadoPagoPosRequest(self.sudo().mp_bearer_token)
+        mercado_pago = MercadoPagoPosRequest(self.sudo().pos_payment_provider_id.mp_bearer_token)
         # Call Mercado Pago for payment intend cancelation
         resp = mercado_pago.call_mercado_pago("delete", f"/point/integration-api/devices/{self.mp_id_point_smart_complet}/payment-intents/{payment_intent_id}", {})
         _logger.debug("mp_payment_intent_cancel(), response from Mercado Pago: %s", resp)
         return resp
 
+    def write(self, vals):
+        records = super().write(vals)
+
```
Comment (paso-odoo) [inline addons/pos_mercado_pago/models/pos_payment_method.py:60]:
```suggestion
```

File: addons/pos_mercado_pago/models/pos_payment_method.py [hunk 2]
```diff
@@ -78,12 +49,20 @@ def mp_payment_intent_cancel(self, payment_intent_id):
...
+        if 'pos_payment_provider_id' in vals and self.use_payment_terminal == 'mercado_pago':
+            self.mp_id_point_smart_complet = self._find_terminal(self.pos_payment_provider_id.mp_bearer_token, self.terminal_id)
+
```
Comment (paso-odoo) [inline addons/pos_mercado_pago/models/pos_payment_method.py:63]:
```suggestion
```

File: addons/pos_online_payment/models/pos_payment_method.py [hunk 3]
```diff
@@ -97,7 +97,12 @@ def _force_online_payment_values(vals, if_present=False):
             vals['payment_method_type'] = 'none'
 
     def _get_payment_terminal_selection(self):
-        return super(PosPaymentMethod, self)._get_payment_terminal_selection() if not self.is_online_payment else []
+        selection = []
+        if len(self.ids) == 0:
+            return super()._get_payment_terminal_selection() if not self.is_online_payment else []
+        for method in self:
+            selection.append(super(PosPaymentMethod, method)._get_payment_terminal_selection() if not method.is_online_payment else [])
+        return selection
```
Comment (paso-odoo) [inline addons/pos_online_payment/models/pos_payment_method.py:105]:
```suggestion
        terminal_selections = []
        for method in self:
            if not method.is_online_payment:
                terminal_selections.append(super(PosPaymentMethod, self)._get_payment_terminal_selection())
        else:
            return super()._get_payment_terminal_selection() if not self.is_online_payment else []    
        return terminal_selections
```

I would do something like this

File: addons/pos_paytm/models/pos_payment_provider.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+from odoo import fields, models
+
+
+class PosPaymentProvider(models.Model):
+    _inherit = "pos.payment.provider"
+
+    code = fields.Selection(
+        selection_add=[('paytm', "PayTM")], ondelete={'paytm': 'set default'})
```
Comment (paso-odoo) [inline addons/pos_paytm/models/pos_payment_provider.py:8]:
```suggestion
        selection_add=[('paytm', 'PayTM')], ondelete={'paytm': 'set default'})
```

File: addons/pos_paytm/models/pos_payment_provider.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    channel_id = fields.Char(string='PayTM Channel ID', default='EDC', copy=False)
+    accept_payment = fields.Selection(selection=[('auto', 'Automatically'), ('manual', 'Manually')], default='auto', help="Choose accept payment mode: \n Manually or Automatically", copy=False)
+    allowed_payment_modes = fields.Selection(selection=[('all', 'All'), ('card', 'Card'), ('qr', 'QR')], default='all', help="Choose allow payment mode: \n All/Card or QR", copy=False)
+    paytm_mid = fields.Char(string="PayTM Merchant ID", help="Go to https://business.paytm.com/ and create the merchant account", copy=False)
+    paytm_merchant_key = fields.Char(string="PayTM Merchant API Key", help="Merchant/AES key \n ex: B1o6Ivjy8L1@abc9", copy=False)
```
Comment (paso-odoo) [inline addons/pos_paytm/models/pos_payment_provider.py:13]:
```suggestion
    paytm_mid = fields.Char(string='PayTM Merchant ID', help='Go to https://business.paytm.com/ and create the merchant account', copy=False)
    paytm_merchant_key = fields.Char(string='PayTM Merchant API Key', help='Merchant/AES key \n ex: B1o6Ivjy8L1@abc9', copy=False)
```

File: addons/pos_razorpay/models/pos_payment_method.py [hunk 6]
```diff
@@ -1,22 +1,16 @@
...
-    razorpay_username = fields.Char(string='Razorpay Username', help='Username(Device Login) \n ex: 1234500121')
-    razorpay_api_key = fields.Char(string='Razorpay API Key', help='Used when connecting to Razorpay: https://razorpay.com/docs/payments/dashboard/account-settings/api-keys/')
-    razorpay_test_mode = fields.Boolean(string='Razorpay Test Mode', default=False, help='Turn it on when in Test Mode')
-
     def _get_payment_terminal_selection(self):
         return super()._get_payment_terminal_selection() + [('razorpay', 'Razorpay')]
 
     def razorpay_make_payment_request(self, data):
-        razorpay = RazorpayPosRequest(self)
+        razorpay = RazorpayPosRequest(self.pos_payment_provider_id, self.terminal_id)
```
Comment (paso-odoo) [inline addons/pos_razorpay/models/pos_payment_method.py:13]: I would pass self and use other fields from self.
i.e. `razorpay = RazorpayPosRequest(self)`

File: addons/pos_razorpay/models/razorpay_pos_request.py [hunk 7]
```diff
@@ -7,17 +7,17 @@
...
+        self.razorpay_api_key = payment_provider.razorpay_api_key
+        self.razorpay_username = payment_provider.razorpay_username
+        self.terminal_id = terminal_id
+        self.razorpay_allowed_payment_modes = payment_provider.razorpay_allowed_payment_modes
+        self.payment_provider = payment_provider
         self.session = requests.Session()
 
     def _razorpay_get_endpoint(self):
-        if self.razorpay_test_mode:
+        if self.razorpay_test_mode == 'test':
```
Comment (paso-odoo) [inline addons/pos_razorpay/models/razorpay_pos_request.py:20]:
```suggestion
        if self.payment_provider.state == 'test':
```
??

File: addons/pos_razorpay/models/razorpay_pos_request.py [hunk 8]
```diff
@@ -7,17 +7,17 @@
...
-        self.razorpay_tid = payment_method.razorpay_tid
-        self.razorpay_allowed_payment_modes = payment_method.razorpay_allowed_payment_modes
-        self.payment_method = payment_method
+    def __init__(self, payment_provider, terminal_id=None):
+        self.razorpay_test_mode = payment_provider.state
+        self.razorpay_api_key = payment_provider.razorpay_api_key
+        self.razorpay_username = payment_provider.razorpay_username
+        self.terminal_id = terminal_id
+        self.razorpay_allowed_payment_modes = payment_provider.razorpay_allowed_payment_modes
+        self.payment_provider = payment_provider
```
Comment (paso-odoo) [inline addons/pos_razorpay/models/razorpay_pos_request.py:16]:
```suggestion
    def __init__(self, payment_method):
        self.razorpay_test_mode = payment_method.pos_payment_provider_id.state
        self.razorpay_api_key = payment_method.pos_payment_provider_id.razorpay_api_key
        self.razorpay_username = payment_method.pos_payment_provider_id.razorpay_username
        self.terminal_id = payment_method.terminal_id
        self.razorpay_allowed_payment_modes = payment_method.pos_payment_provider_id.razorpay_allowed_payment_modes
        self.payment_provider = payment_method.pos_payment_provider_id
```

**Review by paso-odoo — CHANGES_REQUESTED** [review]

File: addons/point_of_sale/models/pos_payment_method.py [hunk 9]
```diff
@@ -61,14 +61,16 @@ def _get_payment_method_type(self):
         help='Type of QR-code to be generated for this payment method.',
     )
     hide_qr_code_method = fields.Boolean(compute='_compute_hide_qr_code_method')
+    terminal_id = fields.Text()
```
Comment (caburj) [inline addons/point_of_sale/models/pos_payment_method.py:64]:
```suggestion
    terminal_identifier = fields.Text()
```

It's better to reserve the `_id` to a related m2o field.

Comment (caburj) [reply addons/point_of_sale/models/pos_payment_method.py:64]: or `terminal_name`

File: addons/point_of_sale/data/pos_payment_provider_data.xml [hunk 10]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo noupdate="1">
```
Comment (caburj) [inline addons/point_of_sale/data/pos_payment_provider_data.xml:2]: why noupdate?

**Review by paso-odoo — CHANGES_REQUESTED** [review]

File: addons/pos_paytm/models/pos_payment_provider.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    accept_payment = fields.Selection([
+        ('auto', 'Automatically'), ('manual', 'Manually')],
+        default='auto',
+        help='Choose accept payment mode: \n Manually or Automatically',
+        copy=False)
+    allowed_payment_modes = fields.Selection([
+        ('all', 'All'), ('card', 'Card'), ('qr', 'QR')],
+        default='all',
+        help='Choose allow payment mode: \n All/Card or QR',
+        copy=False)
```
Comment (caburj) [inline addons/pos_paytm/models/pos_payment_provider.py:18]: you should probably prefix (with `paytm_`) these fields too.

Comment (caburj) [inline addons/pos_paytm/models/pos_payment_provider.py:18]: you should probably prefix (with `paytm_`) these fields too.

File: addons/pos_adyen/views/pos_payment_provider_views.xml [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="pos_payment_provider_view_form_inherit_pos_adyen" model="ir.ui.view">
+        <field name="name">pos.payment.provider.form.inherit.adyen</field>
+        <field name="model">pos.payment.provider</field>
+        <field name="inherit_id" ref="point_of_sale.pos_payment_provider_view_form"/>
+        <field name="arch" type="xml">
+            <group name="provider_credentials" position='inside'>
+                <group name="adyen_credentials" invisible="code != 'adyen'">
+                       <field name="adyen_api_key" required="code == 'adyen' and state != 'disabled'" password="True"/>
```
Comment (caburj) [inline addons/pos_adyen/views/pos_payment_provider_views.xml:10]: for other keys in other payment providers, maybe it's a good idea to make them `password="True"` just like we do here in `pos_adyen`.

File: addons/point_of_sale/models/pos_payment_provider.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    name = fields.Char(string='Name', required=True, translate=True)
+    code = fields.Selection([
+        ('none', 'No Provider Set')
+        ],
+        string='Code',
+        help='The technical code of this payment provider(terminal).',
+        default='none',
+        required=True,
+    )
+    state = fields.Selection([
```
Comment (caburj) [inline addons/point_of_sale/models/pos_payment_provider.py:18]:
```suggestion
    mode = fields.Selection([
```

Opinion: I think `mode` is a more proper name for this.

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.js [hunk 14]
```diff
@@ -38,7 +38,17 @@ export class PaymentScreen extends Component {
...
+            .filter((pm) => {
+                if (
+                    (pm.terminal_id && !pm.pos_payment_provider_id) ||
+                    (pm.pos_payment_provider_id && pm.pos_payment_provider_id.state === "disabled")
+                ) {
+                    return false;
+                } else {
+                    return true;
+                }
+            });
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.js:51]: how about not loading the payment method to the client app when the linked provider is disabled? That way, you don't need to filter it here in the client app.

Comment (caburj) [reply addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.js:51]: To do this, go to the `pos.payment.method` model and modify the "loading" domain.

File: addons/pos_mercado_pago/models/pos_payment_method.py [hunk 15]
```diff
@@ -78,12 +67,18 @@ def mp_payment_intent_cancel(self, payment_intent_id):
...
+        mercado_pago = MercadoPagoPosRequest(self.sudo().pos_payment_provider_id.mp_bearer_token)
         # Call Mercado Pago for payment intend cancelation
         resp = mercado_pago.call_mercado_pago("delete", f"/point/integration-api/devices/{self.mp_id_point_smart_complet}/payment-intents/{payment_intent_id}", {})
         _logger.debug("mp_payment_intent_cancel(), response from Mercado Pago: %s", resp)
         return resp
 
+    def write(self, vals):
+        records = super().write(vals)
+        if 'pos_payment_provider_id' in vals and self.use_payment_terminal == 'mercado_pago':
+            self.mp_id_point_smart_complet = self._find_terminal(self.pos_payment_provider_id.mp_bearer_token, self.terminal_id)
```
Comment (caburj) [inline addons/pos_mercado_pago/models/pos_payment_method.py:79]:
```suggestion
            self.mp_id_point_smart_complet = self._find_terminal(self.pos_payment_provider_id.mp_bearer_token, self.terminal_identifier)
```

I'm not sure if you forgot to adapt this with the terminal_identifier rename.

File: addons/pos_six/views/pos_payment_method_views.xml [hunk 16]
```diff
@@ -1,11 +1,11 @@
...
 <odoo>
-    <record id="pos_payment_method_view_form_inherit_pos_adyen" model="ir.ui.view">
-        <field name="name">pos.payment.method.form.inherit.adyen</field>
+    <record id="pos_payment_method_view_form_inherit_pos_six" model="ir.ui.view">
+        <field name="name">pos.payment.method.form.inherit.six</field>
         <field name="model">pos.payment.method</field>
         <field name="inherit_id" ref="point_of_sale.pos_payment_method_view_form"/>
         <field name="arch" type="xml">
-            <xpath expr="//field[@name='use_payment_terminal']" position="after">
+            <xpath expr="//field[@name='terminal_id']" position="after">
```
Comment (caburj) [inline addons/pos_six/views/pos_payment_method_views.xml:8]:
```suggestion
            <xpath expr="//field[@name='terminal_identifier']" position="after">
```

File: addons/pos_self_order_adyen/models/pos_payment_method.py [hunk 17]
```diff
@@ -30,7 +30,7 @@ def _payment_request_from_kiosk(self, order):
                         'MessageCategory': "Payment",
                         'SaleID': f'{pos_config.display_name} (ID:{pos_config.id})', #	Your unique ID for the POS system component to send this request from.
                         'ServiceID': str(random_number), # Your unique ID for this request, consisting of 1-10 alphanumeric characters.
-                        'POIID': self.adyen_terminal_identifier, #	The unique ID of the terminal to send this request to.
+                        'POIID': self.terminal_id,  # The unique ID of the terminal to send this request to.
```
[the remark this answers is not in the document]
Comment (paan-odoo) [reply addons/pos_self_order_adyen/models/pos_payment_method.py:33]: @caburj
I am working on it.

File: addons/point_of_sale/models/pos_payment_method.py (hunk 9 above)
Comment (caburj) [reply addons/point_of_sale/models/pos_payment_method.py]: I see that you are now using `terminal_identifier`, however, I think you forgot to change other references.

**Comment by paan-odoo** [comment]
@caburj 
I am still working on changes.

File: addons/pos_stripe/models/pos_payment_provider.py [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+
+class PosPaymentProvider(models.Model):
+    _inherit = 'pos.payment.provider'
+
+    code = fields.Selection(selection_add=[('stripe', 'Stripe')], ondelete={'stripe': 'set default'})
+    is_stripe_provider_configured = fields.Boolean(compute="_compute_is_stripe_provider_configured")
+
+    def _compute_is_stripe_provider_configured(self):
+        for record in self:
+            res_id = self.env['payment.provider'].search([('code', '=', 'stripe'), ('company_id', '=', self.env.company.id), ('stripe_secret_key', '!=', False)], limit=1).id
```
Comment (paso-odoo) [inline addons/pos_stripe/models/pos_payment_provider.py:12]:
```suggestion
            provider_id = self.env['payment.provider'].search([('code', '=', 'stripe'), ('company_id', '=', self.env.company.id), ('stripe_secret_key', '!=', False)], limit=1).id
```

**Review by paso-odoo — CHANGES_REQUESTED** [review]

**Comment by paan-odoo** [comment]
Hello @caburj,

Changes are done.

Thanks!

**Comment by paan-odoo** [comment]
Change are done in [PR](https://github.com/odoo/odoo/pull/175119)

**Pull request closed without merging**